src/category_theory/limits/shapes/images.lean

```diff
@@ -273,3 +274,112 @@ end
 end
 
 end category_theory.limits
+
+namespace category_theory.limits
+
+variables {C : Type u} [𝒞 : category.{v} C]
+include 𝒞
+
+section
+
+instance {X Y : C} (f : X ⟶ Y) [has_image f] : has_image (arrow.mk f).hom :=
+show has_image f, by apply_instance
+
+end
+
+section has_image_map
+
+/-- An image map is a morphism `image f → image g` fitting into a commutative square and satisfying
+    the obvious commutativity conditions. -/
+class has_image_map {f g : arrow C} [has_image f.hom] [has_image g.hom] (sq : f ⟶ g) :=
+(map : image f.hom ⟶ image g.hom)
+(factor_map' : factor_thru_image f.hom ≫ map = sq.left ≫ factor_thru_image g.hom . obviously)
+(map_ι' : map ≫ image.ι g.hom = image.ι f.hom ≫ sq.right . obviously)
+
+restate_axiom has_image_map.factor_map'
+restate_axiom has_image_map.map_ι'
+attribute [simp, reassoc] has_image_map.factor_map has_image_map.map_ι
+
+variables {f g : arrow C} [has_image f.hom] [has_image g.hom] (sq : f ⟶ g)
+
+section
+local attribute [ext] has_image_map
+
+instance : subsingleton (has_image_map sq) :=
+subsingleton.intro $ λ a b, has_image_map.ext a b $ (cancel_mono (image.ι g.hom)).1 $
+  by simp only [has_image_map.map_ι]
+
+end
+
+variable [has_image_map sq]
+
+/-- The map on images induced by a commutative square. -/
+abbreviation image.map : image f.hom ⟶ image g.hom := has_image_map.map sq
+
+lemma image.factor_map :
+  factor_thru_image f.hom ≫ image.map sq = sq.left ≫ factor_thru_image g.hom :=
+by simp
+lemma image.map_ι : image.map sq ≫ image.ι g.hom = image.ι f.hom ≫ sq.right :=
+by simp
+lemma image.map_hom_mk'_ι {X Y P Q : C} {k : X ⟶ Y} [has_image k] {l : P ⟶ Q} [has_image l]
+  {m : X ⟶ P} {n : Y ⟶ Q} (w : m ≫ l = k ≫ n) [has_image_map (arrow.hom_mk' w)] :
+  image.map (arrow.hom_mk' w) ≫ image.ι l = image.ι k ≫ n :=
+image.map_ι _
+
+section
+variables {h : arrow C} [has_image h.hom] (sq' : g ⟶ h)
+variables [has_image_map sq']
+
+/-- Image maps for composable commutative squares induce an image map in the composite square. -/
+def has_image_map_comp : has_image_map (sq ≫ sq') :=
+{ map := image.map sq ≫ image.map sq' }
+
+-- This cannot be a simp lemma, see https://github.com/leanprover-community/lean/issues/181.
+lemma image.map_comp [has_image_map (sq ≫ sq')] :
+  image.map (sq ≫ sq') = image.map sq ≫ image.map sq' :=
+show (has_image_map.map (sq ≫ sq')) = (has_image_map_comp sq sq').map, by congr
+
+end
+
+section
+variables (f)
+
+/-- The identity `image f ⟶ image f` fits into the commutative square represented by the identity
+    morphism `𝟙 f` in the arrow category. -/
+def has_image_map_id : has_image_map (𝟙 f) :=
+{ map := 𝟙 (image f.hom),
+  factor_map' := by erw [arrow.id_left, category.id_comp, category.comp_id],
+  map_ι' := by erw [arrow.id_right, category.id_comp, category.comp_id] }
```

Comment:
```suggestion
{ map := 𝟙 (image f.hom), }
```
Both of these proofs actually work by `simp, dsimp, simp`.

1. It would be good to understand why we need to drop back to `dsimp` here. Unfortunately this seems to be frequently necessary when using highly dependent types.
2. I know that I've written a great many `erw`s in the category_theory library, but most of them are code smells, and we should be suspicious of them.
3. In this case let's leave the proofs out, as they are boring and `tidy` can cope just fine.

Reply:
I haven't been able to fully pin this down, but I think the reason why the `dsimp` is needed here has something to do with the fact that implicit arguments are in the wrong from: `category.id_comp` expects that the goal contains `𝟙 X ≫ f` with `f : X ⟶ Y`, but in reality the codomain of `f` is something like `(𝟭 C).obj X` (actually, in this particular case there's some `auto_param` in there too because of how `comma` is defined, but fixing that doesn't solve the problem) and in non-trivial cases `simp` and `rw category.id_comp` cannot make sense of that and `rw functor.id_obj` produces `motive is not type correct`, while `dsimp only [functor.id_obj]` somehow manages to do the right thing.

Reply:
@gebner provided a [very interesting solution](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/.60dsimp.2C.20simp.60) to this `dsimp, simp` problem!

But for now everything is great here.